test/ruby/test_thread.rb

```diff
@@ -1557,4 +1557,37 @@ def test_pending_interrupt?
     assert_equal(true, t.pending_interrupt?(Exception))
     assert_equal(false, t.pending_interrupt?(ArgumentError))
   end
+
+  def test_deadlock_backtrace
+    bug21127 = '[ruby-core:120930] [Bug #21127]'
+
+
+    expected_stderr = [
+      /-:12:in 'Thread#join': No live threads left. Deadlock\? \(fatal\)\n/,
+      /2 threads, 2 sleeps current:\w+ main thread:\w+\n/,
+      /\* #<Thread:\w+ sleep_forever>\n/,
+      :*,
+      /^\s*-:6:in 'Object#frame_for_deadlock_test_2'/,
+      :*,
+      /\* #<Thread:\w+ -:10 sleep_forever>\n/,
+      :*,
+      /^\s*-:2:in 'Object#frame_for_deadlock_test_1'/,
```

Comment:
It depends on the order of `threads.set`.
https://github.com/pocke/ruby/blob/85fb7a31942884b5b0a0d95a2c25eb01dbfd6e15/thread.c#L5580

I'm concerned it causes a flaky test if the order is randomized. I guess it's not a problem because the main thread is always the first item on the list. Please tell me if it will cause a problem 🙏

Reply:
I'm not sure if the order is guaranteed or deterministic.  But if it passes CI, it should be OK.  If it turns out to cause a test failure, we can always just the test so it works with either order.